text/0000-metric-label-set.md

```diff
@@ -0,0 +1,129 @@
+# Metric `LabelSet` specification
+
+**Status:** `proposed`
+
+Introduce a first-class `LabelSet` API type as a handle on a pre-defined set of labels.  
+
+## Motivation
+
+Labels are key-value pairs used across OpenTelemetry for categorizing spans (attributes, event fields) and metrics (labels).  Treatment of labels in the metrics API is especially important because of certain optimizations that take advantage of pre-defined labels.
+
+This optimization currently applies to metric instrument handles, the result of `GetHandle(Instrument, { Key : Value, ... })` on a metric instrument. By allowing the SDK to pre-compute information about the pair of `(Instrument, { Key : Value, ... })`, this reasoning goes, the metrics SDK has the opportunity make individual `Add()` and `Set()` operations on handles as fast as a few machine instructions.
+
+Adopting a first-class `LabelSet` extends the potential for optimization, significantly, for cases where a the `{ Key : Value, ... }` can be reused despite not re-using  `(Instrument, { Key : Value, ... })` pairs.
+
+The current API specifies one way to use set of labels for more than one measurement,
+
+```
+RecordBatch({ Key1: Value1,
+  	      Key2: Value2,
+	      ... },
+	    [ (Gauge, Value), 
+	      (Cumulative, Value),
+	      (Measure, Value),
+	      ... ])
+```
+
+This RFC proposes the new `LabelSet` concept is returned by an API named `Meter.DefineLabels({ Key: Value, ... })` which allows the SDK to potentially sort, canonicalize, or hash the set of labels once, allowing it to be re-used in several ways.  For example, the `RecordBatch` above can be written as one call to `DefineLabels` and inidividual operations.
+
+```
+Labels := meter.DefineLabels({ Key1: Value1,
+       	  		       Key2: Value2,
+			       ... })
+Cumulative.Add(Value, Labels)
+Gauge.Set(Value, Labels)
+Measure.Record(Value, Labels)
+...
+```
+
+With a first-class `LabelSet`, labels can even be re-used across multiple calls to `RecordBatch`.
+
+## Explanation
+
+Metric instrument APIs which presently take labels in the form `{ Key: Value, ... }` will be updated to take an explicit `LabelSet`.  The `Meter.DefineLabels` API method supports getting a `LabelSet` from the SDK, allowing the programmer to pre-define labels without being required to manage handles.  This brings the number of ways to update a metric to three, via re-using a handle from `GetHandle()`:
+
+```
+cumulative := metric.NewFloat64Cumulative("my_counter", [
+					    "required_key1",
+					    "required_key2",
+					  ])
+labels := meter.DefineLabels({ "required_key1": value1,
+			       "required_key2": value2 })
+handle := cumulative.GetHandle(labels)
```

Comment:
With the introduction of labels as key-value-pairs, we don't have to support the implicit ordering of the keys in the instrument and values in the handle and match them correct?

Reply:
I wrote that it should be optional in `Ordered LabelSet Option` below. I don't think it's important, but some do.